### PR TITLE
Fixes for Julia 0.7

### PR DESCRIPTION
**Pull request by martinholters**
This should fix (most of) the deprecations on Julia 0.7 and make it usable there again. I've only tried `wavplay` on Linux, though, so no idea whether audioqueue works. Some more remarks follow inline.

The tests still show a bunch of deprecation warnings for implicit assignment to global variables. From a cursory glance, these should be safe to ignore because the assigned values are not used in global scope. Best way to get rid of the warnings would probably be wrapping things in testsets (to introduce explicit local scopes).

### CODE REVIEW

File: src/WAV.jl [hunk 1]
```diff
@@ -6,11 +6,13 @@ export WAVChunk, WAVMarker, wav_cue_read, wav_cue_write, wav_info_write, wav_inf
...
 export isextensible, isformat, bits_per_sample
 export WAVE_FORMAT_PCM, WAVE_FORMAT_IEEE_FLOAT, WAVE_FORMAT_ALAW, WAVE_FORMAT_MULAW
+using Compat: codeunits, findall, nothing, Nothing, undef
+import Compat: Libdl
 using FileIO
 
 function __init__()
     module_dir = dirname(@__FILE__)
-    if Libdl.find_library(["libpulse-simple"]) != ""
+    if Libdl.find_library(["libpulse-simple.so.0"]) != ""
```
Comment (martinholters) [inline src/WAV.jl]: Are there systems where this does not work? I.e. should this be `Libdl.find_library(["libpulse-simple.so.0", "libpulse-simple.so"])` or similar?

File: src/WAV.jl [hunk 2]
```diff
@@ -517,7 +519,7 @@ function read_data(io::IO, chunk_size, fmt::WAVFormat, format, subrange)
     # "format" is the format of values, while "fmt" is the WAV file level format
     convert_to_double = x -> convert(Array{Float64}, x)
 
-    if subrange === Void
+    if subrange === nothing || subrange === Nothing # Nothing kept for legacy reasons only
```
Comment (martinholters) [inline src/WAV.jl]: I've changed the default value from the type `Void` (no `Nothing`) to its singleton value `nothing`, which is much more idiomatic. In case anyone explicitly used `subrange=Void`, this will be handled here. Should there be a `depwarn` instead of silently accepting?

File: src/WAV.jl [hunk 3]
```diff
@@ -761,18 +763,18 @@ function wavappend(samples::AbstractArray, io::IO)
...
+    # Compute data length of current chunk to-be-appended.
     data_length = size(samples, 1) * fmt.block_align
-    # Update `chunksize`: add length of new data. 
+    # Update `chunksize`: add length of new data.
     seek(io,4)
     write_le(io, convert(UInt32, chunk_size + data_length))
-    # Get `subchunk2size`. 
-    seek(io,40)
+    # Get `subchunk2size`.
+    seek(io, 24 + subchunk_size)
```
Comment (martinholters) [inline src/WAV.jl]: This seems totally unrelated to the migration from Julia 0.6 to 0.7. How did this ever work with seeking to `40`? Usually the position should be `64`.

File: src/WAVChunk.jl [hunk 4]
```diff
@@ -1,15 +1,15 @@
 """
 A RIFF chunk.
 """
-type WAVChunk
+mutable struct WAVChunk
```
Comment (martinholters) [inline src/WAVChunk.jl]: This could probably be immutable.

File: src/WAV.jl (hunk 1 above)
Comment (dancasimiro) [reply src/WAV.jl]: I don’t think that this works on Mac OS X, where the shared object extension is dylib.

Comment (martinholters) [reply src/WAV.jl]: Ah, I wasn't aware this was used on OSX, too. Does it still work unversioned there? Or would it have to be `"libpulse-simple.0.dylib"` there? Anyway, I can definitely put the unversioned entry back, than at least it won't break anything.

Comment (dancasimiro) [reply src/WAV.jl]: I have not tested recently, but the the bare string used to work on Linux and OS X. I expect it to work on Windows as well, but I have never tested on that operating system.

File: src/WAV.jl (hunk 2 above)
Comment (dancasimiro) [reply src/WAV.jl]: I think that you can break the API here. There is another breaking change on master that will be released with Julia 1.0.

File: src/WAV.jl (hunk 3 above)
Comment (dancasimiro) [reply src/WAV.jl]: This was submitted recently. I think that there is a unit test that goes with it.

File: src/WAVChunk.jl (hunk 4 above)
Comment (dancasimiro) [reply src/WAVChunk.jl]: Is there any code that modifies the type?

File: src/WAV.jl (hunk 3 above)
Comment (dancasimiro) [reply src/WAV.jl]: See pull request #61.

File: src/WAV.jl (hunk 1 above)
Comment (martinholters) [reply src/WAV.jl]: The unversioned string fails on Linux since https://github.com/JuliaLang/julia/pull/22828 (0.7.0-DEV.3588).

File: src/WAV.jl (hunk 2 above)
Comment (martinholters) [reply src/WAV.jl]: > I think that you can break the API here.

Ok. With a deprecation or hard-breaking? Also, on second thought, `subrange=:` might be even better than `subrange=nothing`. Preferences?

> There is another breaking change on master that will be released with Julia 1.0.

But you do plan on releasing a version that works with Julia 0.7 before 1.0 comes out, right?

File: src/WAV.jl (hunk 3 above)
Comment (martinholters) [reply src/WAV.jl]: Ah, yes, without this change, I was seeing the same test failures as in the CI log of that PR:
```
Test Failed
  Expression: length(y) == 2 * length(x)
   Evaluated: 8000 == 16000
ERROR: LoadError: There was an error during testing
while loading /home/travis/.julia/v0.6/WAV/test/runtests.jl, in expression starting on line 45
```
So apparently, #61 fixed a bug and added a test for it, but broke another test. With this update, they both pass. Without investigating closer, I guess the two tests append to WAV files with fmt chunks of different lengths, so a fixed value, 40 or 64, cannot be right for both.

File: src/WAV.jl (hunk 2 above)
Comment (dancasimiro) [reply src/WAV.jl]: Yeah, I do plan to make a release that supports both. Do the Julia deve still plan to make 0.7 api compatible with 1.0? I haven’t been following as closely as I once did.

File: src/WAVChunk.jl (hunk 4 above)
Comment (dancasimiro) [reply src/WAVChunk.jl]: I need to check, but I believe that this type is new. I think that it came with #61 as well, which has not been part of a release.

File: src/WAV.jl (hunk 2 above)
Comment (martinholters) [reply src/WAV.jl]: Yes, 1.0 should only drop functionality that's already deprecated in 0.7. So with this PR, WAV.jl should be ready for Julia 1.0 (fingers crossed).

Meanwhile, what's your opinion on `subrange=:` vs. `subrange=nothing` and deprecation yes/no?

File: src/WAVChunk.jl (hunk 4 above)
Comment (martinholters) [reply src/WAVChunk.jl]: Came with #56, which is also not yet part of a release. So making it immutable now should be safe. If the need arises in the future, one can always make it mutable, while the opposite direction would be breaking. Should I use this opportunity then to make it immutable?

File: src/WAV.jl (hunk 2 above)
Comment (dancasimiro) [reply src/WAV.jl]: I prefer `subrange=:`. You can add a deprecation if it isn't too much of a hassle. I would keep the deprecation for the Julia 0.7 release cycle and remove it when Julia 1.0 ships.

File: src/WAVChunk.jl (hunk 4 above)
Comment (dancasimiro) [reply src/WAVChunk.jl]: Yes, please make the type immutable.
